File: syllabus-1/day-5/FIRESC.py
import sys
sys.setrecursionlimit(10**6)


def dfs(node, arr, visited, cost):
    if(visited[node] == 1):
        return
    visited[node] = 1
    cost[0] += 1
    for i in arr[node]:
        if(visited[i] == 0):
            dfs(i, arr, visited, cost)


def solve():
    (n, e) = map(int, input().split())

    arr = [[] for i in range(n)]
    for i in range(e):
        (a, b) = map(int, input().split())
        a -= 1
        b -= 1
        arr[a].append(b)
        arr[b].append(a)

    res = []
    visited = [0 for i in range(n)]
    for i in range(n):
        cost = [0]
        dfs(i, arr, visited, cost)
        if(cost[0] != 0):
            res.append(cost[0])

    ans = 1
    for i in res:
        ans *= i
        ans %= (int(1e9) + 7)
    print(str(len(res)) + " " + str(ans))


# testcases
t = int(input())
while(t > 0):
    solve()
    t -= 1

File: syllabus-1/day-5/FIRESC.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl "\n"
#define ll long long
#define MOD (ll)1e9+7

void dfs(int node, vector<vector<int>>& arr, vector<int>& visited, int& cost) {
    if (visited[node]) return;

    visited[node] = 1;
    cost++;
    for (int i : arr[node]) {
        if (visited[i] == 0) {
            dfs(i, arr, visited, cost);
        }
    }
}

void solve() {
    int n, e;
    cin >> n >> e;

    vector<vector<int>> arr(n);
    for (int i = 0;i < e;++i) {
        int a, b;
        cin >> a >> b;
        a--;
        b--;

        arr[a].push_back(b);
        arr[b].push_back(a);
    }

    vector<int> res, visited(n, 0);
    for (int i = 0;i < n;++i) {
        int cost = 0;
        dfs(i, arr, visited, cost);

        if (cost != 0) {
            res.push_back(cost);
        }
    }

    ll ans = 1;
    for (int i : res) {
        ans *= i;
        ans %= MOD;
    }
    cout << res.size() << " " << ans << endl;
}

int main() {

    int t = 1;
    cin >> t;
    while (t--) {
        solve();
    }

    return 0;
}